### PR TITLE
[78] Choose the arrangement of newly created elements

### PR DESCRIPTION
**Pull request by GlennPlou**
- SiriusLayoutHelper#updateClobberedPosition() has been modified to
override the default behavior that displays newly created synchronized
elements diagonally, and allow elements to be displayed vertically or
horizontally.
- The SiriusDiagramUiPreferencesKeys#PREF_NEWLY_CREATED_ELEMENTS_LAYOUT
preference has been added to allow the user to choose the layout.
- The SiriusLayoutDataManagerImpl#calculateCenterLocation() method has
been enhanced to handle the case where several figures are created at
once, figures are arranged according to user preferences.

Bug: https://github.com/eclipse-sirius/sirius-desktop/issues/78

### CODE REVIEW

**Review by florianbarbin — COMMENTED** [review]
It looks good to me in general but I have a different behvior with the following use case:
* Get the use case attached to the #78 issue.
* Open the gray diagram and add several elements.
* Open the Colored diagram => all elements are created horizontally (no matter the preference value)

**Review by florianbarbin — APPROVED** [review]

File: plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/graphical/figures/SiriusLayoutHelper.java [hunk 1]
```diff
@@ -87,14 +93,91 @@ public Point getReferencePosition(IFigure parent, Viewport viewport, IGraphicalE
...
+     * user preferences and applied to the specified reference point.
+     * 
+     * @param referencePoint
+     *            the reference point to translate
+     * @param figure
+     *            the figure for which to compute the translated point
+     * @param shiftWithGridValue 
+     * @return the translated point based on user preferences
+     */
+    public Point computeTranslatedPoint(Point referencePoint, IFigure figure, boolean shiftWithGridValue) {
```
Comment (lfasani) [inline plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/graphical/figures/SiriusLayoutHelper.java:163]: change shiftWithGridValue to isFirstCreatedElement

File: plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/graphical/figures/SiriusLayoutHelper.java [hunk 2]
```diff
@@ -87,14 +93,91 @@ public Point getReferencePosition(IFigure parent, Viewport viewport, IGraphicalE
...
+
+            return validatePosition(parent, theBounds);
+        }
+        return theBounds.getLocation();
     }
 
     @Override
     public Point updateClobberedPosition(IFigure clobbered, IFigure newlyAddedChild) {
-        Point clobberedPosition = super.updateClobberedPosition(clobbered, newlyAddedChild);
+        Point clobberedPosition = computeTranslatedPoint(determineReferencePoint(clobbered), clobbered, true);
```
Comment (lfasani) [inline plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/graphical/figures/SiriusLayoutHelper.java:128]: document to explain that this method should be called only for the first created element.
(Indeed, we pass true to determineReferencePoint mthod that will imply that the shifting will be done with the grid width)

**Review by lfasani — CHANGES_REQUESTED** [review]

File: plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/graphical/figures/SiriusLayoutHelper.java [hunk 3]
```diff
@@ -87,14 +93,91 @@ public Point getReferencePosition(IFigure parent, Viewport viewport, IGraphicalE
...
+     * @return the translated point based on user preferences
+     */
+    public Point computeTranslatedPoint(Point referencePoint, IFigure figure, boolean shiftWithGridValue) {
+        
+        //si snapToGrid, décaler de la valeur de la grille otherwise we may have a stack overflow in the case two steps of shifting is snaped at the same location
+        // TODO mettre le if
+        int shiftingValue = 30;
         EditPartQuery editPartQuery = new EditPartQuery(this.containerEditPart);
-        return editPartQuery.getSnapLocation(new CreateRequest(), clobberedPosition);
+        if (editPartQuery.isSnapToGridEnabled() && shiftWithGridValue) {
```
[the remark this answers is not in the document]
Comment (GlennPlou) [reply plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/graphical/figures/SiriusLayoutHelper.java:169]: done

File: plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/graphical/figures/SiriusLayoutHelper.java (hunk 1 above)
Comment (GlennPlou) [reply plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/graphical/figures/SiriusLayoutHelper.java:163]: done

File: plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/graphical/figures/SiriusLayoutHelper.java (hunk 2 above)
Comment (GlennPlou) [reply plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/graphical/figures/SiriusLayoutHelper.java:106]: done

**Pull request merged**